packages/aws-cdk/lib/cdk-toolkit.ts

```diff
@@ -295,19 +295,18 @@ export class CdkToolkit {
     // |            |                |      | <------------------ |            | <------------------ |            | <-------------|
     // --------------                --------  'cdk deploy' done  --------------  'cdk deploy' done  --------------
     let latch: 'pre-ready' | 'open' | 'deploying' | 'queued' = 'pre-ready';
-    chokidar.watch(watchIncludes, {
-      ignored: watchExcludes,
-      cwd: rootDir,
-      // ignoreInitial: true,
-    }).on('ready', () => {
-      latch = 'open';
-      debug("'watch' received the 'ready' event. From now on, all file changes will trigger a deployment");
-    }).on('all', async (event, filePath) => {
+
```

Comment:
One thing I think we can improve here is that the extracted function is too big. We need only the deployment part (when `latch` is `'open'`), but we extracted all of it, even the parts that we don't need.

So, I think the way to go is something like this:

```ts
    chokidar.watch(watchIncludes, {
      ignored: watchExcludes,
      cwd: rootDir,
      // ignoreInitial: true,
    }).on('ready', () => {
      latch = 'open';
      debug("'watch' received the 'ready' event. From now on, all file changes will trigger a deployment");
      print("Triggering initial 'cdk deploy'");
      await deployWhileObservingForFileChanges();
    }).on('all', async (event, filePath) => {
      if (latch === 'pre-ready') {
        print(`'watch' is observing ${event === 'addDir' ? 'directory' : 'the file'} '%s' for changes`, filePath);
      } else if (latch === 'open') {
        print("Detected change to '%s' (type: %s). Triggering 'cdk deploy'", filePath, event);
        await deployWhileObservingForFileChanges(); // this name TBD of course
      } else { // this means latch is either 'deploying' or 'queued'
        latch = 'queued';
        print("Detected change to '%s' (type: %s) while 'cdk deploy' is still running. " +
            'Will queue for another deployment after this one finishes', filePath, event);
      }
    });
```

And `deployWhileObservingForFileChanges` is:

```ts
      async function deployWhileObservingForFileChanges() {
        latch = 'deploying';

        await this.invokeDeployFromWatch(options);

        // If latch is still 'deploying' after the 'await', that's fine,
        // but if it's 'queued', that means we need to deploy again
        while ((latch as 'deploying' | 'queued') === 'queued') {
          // TypeScript doesn't realize latch can change between 'awaits',
          // and thinks the above 'while' condition is always 'false' without the cast
          latch = 'deploying';
          print("Detected file changes during deployment. Invoking 'cdk deploy' again");
          await this.invokeDeployFromWatch(options);
        }
        latch = 'open';
      }
```

Reply:
currently named `depoyAndWatch`, hopefully that is reasonable.